[Core][flaky-test] linux://python/ray/tests:test_usage_stats is flaky

### What happened + What you expected to happen

linux://python/ray/tests:test_usage_stats has been flaky with following errors:
```


=================================== FAILURES ===================================
--
  | __________________________ test_usage_lib_report_data __________________________
  |  
  | monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7f591d758898>
  | shutdown_only = None
  | tmp_path = PosixPath('/tmp/pytest-of-root/pytest-3/test_usage_lib_report_data0')
  |  
  | @pytest.mark.skipif(
  | sys.platform == "win32",
  | reason="Test depends on runtime env feature not supported on Windows.",
  | )
  | def test_usage_lib_report_data(monkeypatch, shutdown_only, tmp_path):
  | with monkeypatch.context() as m:
  | m.setenv("RAY_USAGE_STATS_ENABLED", "1")
  | m.setenv("RAY_USAGE_STATS_REPORT_URL", "http://127.0.0.1:8000")
  | # Runtime env is required to run this test in minimal installation test.
  | ray.init(num_cpus=0, runtime_env={"pip": ["ray[serve]"]})
  | """
  | Make sure the generated data is following the schema.
  | """
  | cluster_metadata = ray_usage_lib.get_cluster_metadata(
  | ray.experimental.internal_kv.internal_kv_get_gcs_client(), num_retries=20
  | )
  | cluster_config_file_path = tmp_path / "ray_bootstrap_config.yaml"
  | cluster_config_file_path.write_text(
  | """
  | cluster_name: minimal
  | max_workers: 1
  | provider:
  | type: aws
  | region: us-west-2
  | availability_zone: us-west-2a
  | """
  | )
  | cluster_config_to_report = ray_usage_lib.get_cluster_config_to_report(
  | cluster_config_file_path
  | )
  | d = ray_usage_lib.generate_report_data(
  | cluster_metadata, cluster_config_to_report, 2, 2, 2
  | )
  | validate(instance=asdict(d), schema=schema)
  |  
  | """
  | Make sure writing to a file works as expected
  | """
  | client = ray_usage_lib.UsageReportClient()
  | temp_dir = Path(tmp_path)
  | client.write_usage_data(d, temp_dir)
  |  
  | wait_for_condition(lambda: file_exists(temp_dir))
  |  
  | """
  | Make sure report usage data works as expected
  | """
  |  
  | @ray.remote(num_cpus=0, runtime_env={"pip": ["ray[serve]"]})
  | class ServeInitator:
  | def __init__(self):
  | # Start the ray serve server to verify requests are sent
  | # to the right place.
  | from ray import serve
  |  
  | serve.start()
  |  
  | @serve.deployment(ray_actor_options={"num_cpus": 0})
  | async def usage(request):
  | body = await request.json()
  | if body == asdict(d):
  | return True
  | else:
  | return False
  |  
  | usage.deploy()
  |  
  | def ready(self):
  | pass
  |  
  | # We need to start a serve with runtime env to make this test
  | # work with minimal installation.
  | s = ServeInitator.remote()
  | >           ray.get(s.ready.remote())
  |  
  | /ray/python/ray/tests/test_usage_stats.py:444:
  | _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
  | /ray/python/ray/_private/client_mode_hook.py:105: in wrapper
  | return func(*args, **kwargs)
  | _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
  |  
  | object_refs = [ObjectRef(16310a0f0a45af5cf3019b4336c3332f5b1656030100000001000000)]
  |  
  | @PublicAPI
  | @client_mode_hook(auto_init=True)
  | def get(
  | object_refs: Union[ray.ObjectRef, List[ray.ObjectRef]],
  | *,
  | timeout: Optional[float] = None,
  | ) -> Union[Any, List[Any]]:
  | """Get a remote object or a list of remote objects from the object store.
  |  
  | This method blocks until the object corresponding to the object ref is
  | available in the local object store. If this object is not in the local
  | object store, it will be shipped from an object store that has it (once the
  | object has been created). If object_refs is a list, then the objects
  | corresponding to each object in the list will be returned.
  |  
  | Ordering for an input list of object refs is preserved for each object
  | returned. That is, if an object ref to A precedes an object ref to B in the
  | input list, then A will precede B in the returned list.
  |  
  | This method will issue a warning if it's running inside async context,
  | you can use ``await object_ref`` instead of ``ray.get(object_ref)``. For
  | a list of object refs, you can use ``await asyncio.gather(*object_refs)``.
  |  
  | Args:
  | object_refs: Object ref of the object to get or a list of object refs
  | to get.
  | timeout (Optional[float]): The maximum amount of time in seconds to
  | wait before returning.
  |  
  | Returns:
  | A Python object or a list of Python objects.
  |  
  | Raises:
  | GetTimeoutError: A GetTimeoutError is raised if a timeout is set and
  | the get takes longer than timeout to return.
  | Exception: An exception is raised if the task that created the object
  | or that created one of the objects raised an exception.
  | """
  | worker = global_worker
  | worker.check_connected()
  |  
  | if hasattr(worker, "core_worker") and worker.core_worker.current_actor_is_asyncio():
  | global blocking_get_inside_async_warned
  | if not blocking_get_inside_async_warned:
  | logger.warning(
  | "Using blocking ray.get inside async actor. "
  | "This blocks the event loop. Please use `await` "
  | "on object ref with asyncio.gather if you want to "
  | "yield execution to the event loop instead."
  | )
  | blocking_get_inside_async_warned = True
  |  
  | with profiling.profile("ray.get"):
  | is_individual_id = isinstance(object_refs, ray.ObjectRef)
  | if is_individual_id:
  | object_refs = [object_refs]
  |  
  | if not isinstance(object_refs, list):
  | raise ValueError(
  | "'object_refs' must either be an object ref "
  | "or a list of object refs."
  | )
  |  
  | # TODO(ujvl): Consider how to allow user to retrieve the ready objects.
  | values, debugger_breakpoint = worker.get_objects(object_refs, timeout=timeout)
  | for i, value in enumerate(values):
  | if isinstance(value, RayError):
  | if isinstance(value, ray.exceptions.ObjectLostError):
  | worker.core_worker.dump_object_store_memory_usage()
  | if isinstance(value, RayTaskError):
  | raise value.as_instanceof_cause()
  | else:
  | >                       raise value
  | E                       ray.exceptions.RuntimeEnvSetupError: Failed to setup runtime environment.
  | E                       Could not create the actor because its associated runtime env failed to be created.
  | E                       Traceback (most recent call last):
  | E                         File "/ray/python/ray/dashboard/modules/runtime_env/runtime_env_agent.py", line 388, in _create_runtime_env_with_retry
  | E                           runtime_env_setup_task, timeout=setup_timeout_seconds
  | E                         File "/opt/miniconda/lib/python3.6/asyncio/tasks.py", line 358, in wait_for
  | E                           return fut.result()
  | E                         File "/ray/python/ray/dashboard/modules/runtime_env/runtime_env_agent.py", line 297, in _setup_runtime_env
  | E                           uri, runtime_env, context, logger=per_job_logger
  | E                         File "/ray/python/ray/_private/runtime_env/pip.py", line 440, in create
  | E                           return await task
  | E                         File "/ray/python/ray/_private/runtime_env/pip.py", line 431, in _create_for_hash
  | E                           await PipProcessor(target_dir, runtime_env, logger)
  | E                         File "/ray/python/ray/_private/runtime_env/pip.py", line 331, in _run
  | E                           await self._create_or_get_virtualenv(path, exec_cwd, logger)
  | E                         File "/ray/python/ray/_private/runtime_env/pip.py", line 273, in _create_or_get_virtualenv
  | E                           await check_output_cmd(create_venv_cmd, logger=logger, cwd=cwd, env=env)
  | E                         File "/ray/python/ray/_private/runtime_env/utils.py", line 102, in check_output_cmd
  | E                           proc.returncode, cmd, output=stdout, cmd_index=cmd_index
  | E                       ray._private.runtime_env.utils.SubprocessCalledProcessError: Run cmd[7] failed with the following details.
  | E                       Command '['/opt/miniconda/bin/python3', '-m', 'virtualenv', '--app-data', '/tmp/ray/session_2022-04-15_03-50-51_422041_2132/runtime_resources/pip/5f03e216e7a7205e2f91d630e44c9c6e5040e702/virtualenv_app_data', '--reset-app-data', '--no-periodic-update', '--system-site-packages', '--no-download', '/tmp/ray/session_2022-04-15_03-50-51_422041_2132/runtime_resources/pip/5f03e216e7a7205e2f91d630e44c9c6e5040e702/virtualenv']' returned non-zero exit status 1.
  | E                       Last 50 lines of stdout:
  | E                           RuntimeError: failed to build image setuptools, pip because:
  | E                           Traceback (most recent call last):
  | E                             File "/opt/miniconda/lib/python3.6/site-packages/virtualenv/seed/embed/via_app_data/via_app_data.py", line 57, in _install
  | E                               installer.install(creator.interpreter.version_info)
  | E                             File "/opt/miniconda/lib/python3.6/site-packages/virtualenv/seed/embed/via_app_data/pip_install/base.py", line 35, in install
  | E                               self._uninstall_previous_version()
  | E                             File "/opt/miniconda/lib/python3.6/site-packages/virtualenv/seed/embed/via_app_data/pip_install/base.py", line 151, in _uninstall_previous_version
  | E                               dist_name = self._dist_info.stem.split("-")[0]
  | E                             File "/opt/miniconda/lib/python3.6/site-packages/virtualenv/seed/embed/via_app_data/pip_install/base.py", line 115, in _dist_info
  | E                               raise RuntimeError(msg)  # pragma: no cover
  | E                           RuntimeError: no .dist-info at /tmp/ray/session_2022-04-15_03-50-51_422041_2132/runtime_resources/pip/5f03e216e7a7205e2f91d630e44c9c6e5040e702/virtualenv_app_data/wheel/3.6/image/1/CopyPipInstall/setuptools-59.6.0-py3-none-any, has pkg_resources, setuptools
  | E
  | E                           Traceback (most recent call last):
  | E                             File "/opt/miniconda/lib/python3.6/site-packages/virtualenv/seed/embed/via_app_data/via_app_data.py", line 57, in _install
  | E                               installer.install(creator.interpreter.version_info)
  | E                             File "/opt/miniconda/lib/python3.6/site-packages/virtualenv/seed/embed/via_app_data/pip_install/base.py", line 35, in install
  | E                               self._uninstall_previous_version()
  | E                             File "/opt/miniconda/lib/python3.6/site-packages/virtualenv/seed/embed/via_app_data/pip_install/base.py", line 151, in _uninstall_previous_version
  | E                               dist_name = self._dist_info.stem.split("-")[0]
  | E                             File "/opt/miniconda/lib/python3.6/site-packages/virtualenv/seed/embed/via_app_data/pip_install/base.py", line 115, in _dist_info
  | E                               raise RuntimeError(msg)  # pragma: no cover
  | E                           RuntimeError: no .dist-info at /tmp/ray/session_2022-04-15_03-50-51_422041_2132/runtime_resources/pip/5f03e216e7a7205e2f91d630e44c9c6e5040e702/virtualenv_app_data/wheel/3.6/image/1/CopyPipInstall/pip-21.3.1-py3-none-any, has pip
  |  
  | /ray/python/ray/worker.py:1812: RuntimeEnvSetupError

```

@rkooo567 @jjyao could help debug or triage?

### Versions / Dependencies

latest master

### Reproduction script

check buildkite or flaky test dashboard.

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

--- a/python/ray/tests/test_gcs_fault_tolerance.py
+++ b/python/ray/tests/test_gcs_fault_tolerance.py
@@ -19,6 +19,7 @@ from ray._private.test_utils import (
     run_string_as_driver,
 )
 from ray.job_submission import JobSubmissionClient, JobStatus
+from ray._raylet import GcsClient
 
 import psutil
 
@@ -992,61 +993,60 @@ def test_job_finished_after_head_node_restart(
     cluster = ray_start_cluster_head_with_external_redis
     head_node = cluster.head_node
 
-    client = JobSubmissionClient(head_node.address)
-
     # submit job
-    job_id = client.submit_job(
+    client = JobSubmissionClient(head_node.address)
+    submission_id = client.submit_job(
         entrypoint="python -c 'import ray; ray.init(); print(ray.cluster_resources()); \
-            import time; time.sleep(100)'"
+            import time; time.sleep(1000)'"
     )
 
-    def _check_job_status(client, job_status):
-        return client.get_job_info(job_id).status == job_status
-
-    wait_for_condition(_check_job_status, client=client, job_status=JobStatus.RUNNING)
-
-    # kill and restart head node
-    head_node.kill_raylet()
-    wait_for_condition(
-        lambda: not cluster.global_state.node_table()[0]["Alive"], timeout=30
-    )
-
-    wait_for_condition(_check_job_status, client=client, job_status=JobStatus.FAILED)
-
-    new_head_node = cluster.head_node
-    new_gcs_address = new_head_node.address
-
-    async def async_get_all_job_info():
-        gcs_aio_client = gcs_utils.GcsAioClient(
-            address=new_gcs_address, nums_reconnect_retry=20
-        )
-        return await gcs_aio_client.get_all_job_info()
-
-    def get_job_info(job_id):
-        import asyncio
-
-        asyncio.set_event_loop(asyncio.new_event_loop())
-        all_job_info = get_or_create_event_loop().run_until_complete(
-            async_get_all_job_info()
-        )
+    def get_job_info(submission_id):
+        gcs_client = GcsClient(cluster.address)
+        all_job_info = gcs_client.get_all_job_info()
 
         return list(
             filter(
                 lambda job_info: "job_submission_id" in job_info.config.metadata
-                and job_info.config.metadata["job_submission_id"] == job_id,
+                and job_info.config.metadata["job_submission_id"] == submission_id,
                 list(all_job_info.values()),
             )
         )
 
+    def _check_job_running(client: JobSubmissionClient, submission_id: str) -> bool:
+        job_infos = get_job_info(submission_id)
+        if len(job_infos) == 0:
+            return False
+        job_info = job_infos[0].job_info
+        return job_info.status == JobStatus.RUNNING
+
+    # wait until job info is written in redis
+    wait_for_condition(_check_job_running, client=client, submission_id=submission_id, timeout=10)
+
+    # kill head node
+    ray.shutdown()
+    gcs_server_process = head_node.all_processes["gcs_server"][0].process
+    gcs_server_pid = gcs_server_process.pid
+
+    cluster.remove_node(head_node)
+
+    # Wait to prevent the gcs server process becoming zombie.
+    gcs_server_process.wait()
+    wait_for_pid_to_exit(gcs_server_pid, 1000)
+
+    # restart head node
+    new_head_node = cluster.add_node()
+    ray.init(cluster.address)
+    new_gcs_address = new_head_node.address
+
     # verify if job is finished, which marked is_dead
-    def _check_job_is_dead(job_id: str) -> bool:
-        job_infos = get_job_info(job_id)
+    def _check_job_is_dead(submission_id: str) -> bool:
+        job_infos = get_job_info(submission_id)
         if len(job_infos) == 0:
             return False
         job_info = job_infos[0]
         return job_info.is_dead
 
-    wait_for_condition(_check_job_is_dead, job_id=job_id, timeout=20)
+    wait_for_condition(_check_job_is_dead, submission_id=submission_id, timeout=10)
 
 
 if __name__ == "__main__":

--- a/python/ray/tests/test_gcs_fault_tolerance.py
+++ b/python/ray/tests/test_gcs_fault_tolerance.py
@@ -1038,7 +1038,6 @@ def test_job_finished_after_head_node_restart(
     # restart head node
     new_head_node = cluster.add_node()
     ray.init(cluster.address)
-    new_gcs_address = new_head_node.address
 
     # verify if job is finished, which marked is_dead
     def _check_job_is_dead(submission_id: str) -> bool:

--- a/python/ray/tests/test_gcs_fault_tolerance.py
+++ b/python/ray/tests/test_gcs_fault_tolerance.py
@@ -11,6 +11,7 @@ from ray.util.scheduling_strategies import PlacementGroupSchedulingStrategy
 import ray._private.gcs_utils as gcs_utils
 from ray._private import ray_constants
 from ray._private.test_utils import (
+    format_web_url,
     convert_actor_state,
     enable_external_redis,
     generate_system_config_map,
@@ -18,6 +19,7 @@ from ray._private.test_utils import (
     wait_for_pid_to_exit,
     run_string_as_driver,
 )
+from ray.job_submission import JobStatus, JobSubmissionClient
 
 import psutil
 
@@ -973,6 +975,55 @@ def test_redis_logs(external_redis):
                 "--force",
             ],
         )
+
+
+@pytest.mark.parametrize(
+    "ray_start_regular_with_external_redis",
+    [
+        generate_system_config_map(
+            gcs_failover_worker_reconnect_timeout=20,
+            gcs_rpc_server_reconnect_timeout_s=60,
+        )
+    ],
+    indirect=True,
+)
+def test_job_finished_after_gcs_server_restart(ray_start_regular_with_external_redis):
+    gcs_address = ray._private.worker.global_worker.gcs_client.address
+    client = JobSubmissionClient(gcs_address)
+
+    # submit job
+    job_id = client.submit_job(
+        entrypoint="python -c 'import ray; ray.init(); print(ray.cluster_resources());'",
+    )
+    # restart the gcs server
+    ray._private.worker._global_node.kill_gcs_server()
+    ray._private.worker._global_node.start_gcs_server()
+
+    gcs_address = ray._private.worker.global_worker.gcs_client.address
+
+    async def async_get_all_job_info():
+        gcs_aio_client = gcs_utils.GcsAioClient(
+            address=gcs_address, nums_reconnect_retry=20
+        )
+        return await gcs_aio_client.get_all_job_info()
+
+    def get_job_info(job_id):
+        import asyncio
+
+        asyncio.set_event_loop(asyncio.new_event_loop())
+        all_job_info = get_or_create_event_loop().run_until_complete(async_get_all_job_info())
+        return list(filter(lambda job_info: 'job_submission_id' in job_info.config.metadata \
+            and job_info.config.metadata['job_submission_id'] == job_id, list(all_job_info.values())))
+
+    # verify if job is finished, which marked is_dead
+    def _check_job_is_dead(job_id: str) -> bool:
+        job_infos = get_job_info(job_id)
+        if len(job_infos) == 0:
+            return False
+        job_info = job_infos[0]
+        return job_info.is_dead
+
+    wait_for_condition(_check_job_is_dead, job_id=job_id, timeout=20)
 
 
 if __name__ == "__main__":

--- a/python/ray/tests/test_gcs_fault_tolerance.py
+++ b/python/ray/tests/test_gcs_fault_tolerance.py
@@ -1011,9 +1011,16 @@ def test_job_finished_after_gcs_server_restart(ray_start_regular_with_external_r
         import asyncio
 
         asyncio.set_event_loop(asyncio.new_event_loop())
-        all_job_info = get_or_create_event_loop().run_until_complete(async_get_all_job_info())
-        return list(filter(lambda job_info: 'job_submission_id' in job_info.config.metadata \
-            and job_info.config.metadata['job_submission_id'] == job_id, list(all_job_info.values())))
+        all_job_info = get_or_create_event_loop().run_until_complete(
+            async_get_all_job_info()
+        )
+        return list(
+            filter(
+                lambda job_info: "job_submission_id" in job_info.config.metadata
+                and job_info.config.metadata["job_submission_id"] == job_id,
+                list(all_job_info.values()),
+            )
+        )
 
     # verify if job is finished, which marked is_dead
     def _check_job_is_dead(job_id: str) -> bool:

--- a/python/ray/tests/test_gcs_fault_tolerance.py
+++ b/python/ray/tests/test_gcs_fault_tolerance.py
@@ -1012,7 +1012,7 @@ def test_job_finished_after_head_node_restart(
             )
         )
 
-    def _check_job_running(client: JobSubmissionClient, submission_id: str) -> bool:
+    def _check_job_running(submission_id: str) -> bool:
         job_infos = get_job_info(submission_id)
         if len(job_infos) == 0:
             return False
@@ -1020,9 +1020,7 @@ def test_job_finished_after_head_node_restart(
         return job_info.status == JobStatus.RUNNING
 
     # wait until job info is written in redis
-    wait_for_condition(
-        _check_job_running, client=client, submission_id=submission_id, timeout=10
-    )
+    wait_for_condition(_check_job_running, submission_id=submission_id, timeout=10)
 
     # kill head node
     ray.shutdown()
@@ -1036,7 +1034,7 @@ def test_job_finished_after_head_node_restart(
     wait_for_pid_to_exit(gcs_server_pid, 1000)
 
     # restart head node
-    new_head_node = cluster.add_node()
+    cluster.add_node()
     ray.init(cluster.address)
 
     # verify if job is finished, which marked is_dead

--- a/python/ray/tests/test_gcs_fault_tolerance.py
+++ b/python/ray/tests/test_gcs_fault_tolerance.py
@@ -18,7 +18,7 @@ from ray._private.test_utils import (
     wait_for_pid_to_exit,
     run_string_as_driver,
 )
-from ray.job_submission import JobSubmissionClient
+from ray.job_submission import JobSubmissionClient, JobStatus
 
 import psutil
 
@@ -977,32 +977,48 @@ def test_redis_logs(external_redis):
 
 
 @pytest.mark.parametrize(
-    "ray_start_regular_with_external_redis",
+    "ray_start_cluster_head_with_external_redis",
     [
         generate_system_config_map(
             gcs_failover_worker_reconnect_timeout=20,
-            gcs_rpc_server_reconnect_timeout_s=60,
+            gcs_rpc_server_reconnect_timeout_s=2,
         )
     ],
     indirect=True,
 )
-def test_job_finished_after_gcs_server_restart(ray_start_regular_with_external_redis):
-    gcs_address = ray._private.worker.global_worker.gcs_client.address
-    client = JobSubmissionClient(gcs_address)
+def test_job_finished_after_head_node_restart(
+    ray_start_cluster_head_with_external_redis,
+):
+    cluster = ray_start_cluster_head_with_external_redis
+    head_node = cluster.head_node
+
+    client = JobSubmissionClient(head_node.address)
 
     # submit job
     job_id = client.submit_job(
-        entrypoint="python -c 'import ray; ray.init(); print(ray.cluster_resources());'"
+        entrypoint="python -c 'import ray; ray.init(); print(ray.cluster_resources()); \
+            import time; time.sleep(100)'"
     )
-    # restart the gcs server
-    ray._private.worker._global_node.kill_gcs_server()
-    ray._private.worker._global_node.start_gcs_server()
 
-    gcs_address = ray._private.worker.global_worker.gcs_client.address
+    def _check_job_status(client, job_status):
+        return client.get_job_info(job_id).status == job_status
+
+    wait_for_condition(_check_job_status, client=client, job_status=JobStatus.RUNNING)
+
+    # kill and restart head node
+    head_node.kill_raylet()
+    wait_for_condition(
+        lambda: not cluster.global_state.node_table()[0]["Alive"], timeout=30
+    )
+
+    wait_for_condition(_check_job_status, client=client, job_status=JobStatus.FAILED)
+
+    new_head_node = cluster.head_node
+    new_gcs_address = new_head_node.address
 
     async def async_get_all_job_info():
         gcs_aio_client = gcs_utils.GcsAioClient(
-            address=gcs_address, nums_reconnect_retry=20
+            address=new_gcs_address, nums_reconnect_retry=20
         )
         return await gcs_aio_client.get_all_job_info()
 
@@ -1013,6 +1029,7 @@ def test_job_finished_after_gcs_server_restart(ray_start_regular_with_external_r
         all_job_info = get_or_create_event_loop().run_until_complete(
             async_get_all_job_info()
         )
+
         return list(
             filter(
                 lambda job_info: "job_submission_id" in job_info.config.metadata

--- a/python/ray/tests/test_gcs_fault_tolerance.py
+++ b/python/ray/tests/test_gcs_fault_tolerance.py
@@ -11,7 +11,6 @@ from ray.util.scheduling_strategies import PlacementGroupSchedulingStrategy
 import ray._private.gcs_utils as gcs_utils
 from ray._private import ray_constants
 from ray._private.test_utils import (
-    format_web_url,
     convert_actor_state,
     enable_external_redis,
     generate_system_config_map,
@@ -19,7 +18,7 @@ from ray._private.test_utils import (
     wait_for_pid_to_exit,
     run_string_as_driver,
 )
-from ray.job_submission import JobStatus, JobSubmissionClient
+from ray.job_submission import JobSubmissionClient
 
 import psutil
 
@@ -993,7 +992,7 @@ def test_job_finished_after_gcs_server_restart(ray_start_regular_with_external_r
 
     # submit job
     job_id = client.submit_job(
-        entrypoint="python -c 'import ray; ray.init(); print(ray.cluster_resources());'",
+        entrypoint="python -c 'import ray; ray.init(); print(ray.cluster_resources());'"
     )
     # restart the gcs server
     ray._private.worker._global_node.kill_gcs_server()

--- a/python/ray/tests/test_gcs_fault_tolerance.py
+++ b/python/ray/tests/test_gcs_fault_tolerance.py
@@ -1020,7 +1020,9 @@ def test_job_finished_after_head_node_restart(
         return job_info.status == JobStatus.RUNNING
 
     # wait until job info is written in redis
-    wait_for_condition(_check_job_running, client=client, submission_id=submission_id, timeout=10)
+    wait_for_condition(
+        _check_job_running, client=client, submission_id=submission_id, timeout=10
+    )
 
     # kill head node
     ray.shutdown()

--- a/src/ray/gcs/gcs_server/gcs_job_manager.cc
+++ b/src/ray/gcs/gcs_server/gcs_job_manager.cc
@@ -222,7 +222,7 @@ void GcsJobManager::HandleGetAllJobInfo(rpc::GetAllJobInfoRequest request,
                        << WorkerID::FromBinary(data.second.driver_address().worker_id());
         client->NumPendingTasks(
             std::move(request),
-            [reply, i, num_processed_jobs, try_send_reply](
+            [data, reply, i, num_processed_jobs, try_send_reply](
                 const Status &status,
                 const rpc::NumPendingTasksReply &num_pending_tasks_reply) {
               RAY_LOG(DEBUG) << "SendNumPendingTaskReply: "

--- a/src/ray/gcs/gcs_server/gcs_job_manager.cc
+++ b/src/ray/gcs/gcs_server/gcs_job_manager.cc
@@ -218,13 +218,16 @@ void GcsJobManager::HandleGetAllJobInfo(rpc::GetAllJobInfoRequest request,
         auto client = core_worker_clients_.GetOrConnect(data.second.driver_address());
         std::unique_ptr<rpc::NumPendingTasksRequest> request(
             new rpc::NumPendingTasksRequest());
-        RAY_LOG(DEBUG) << "SendNumPendingTask: " << WorkerID::FromBinary(data.second.driver_address().worker_id());
+        RAY_LOG(DEBUG) << "SendNumPendingTask: "
+                       << WorkerID::FromBinary(data.second.driver_address().worker_id());
         client->NumPendingTasks(
             std::move(request),
             [reply, i, num_processed_jobs, try_send_reply](
                 const Status &status,
                 const rpc::NumPendingTasksReply &num_pending_tasks_reply) {
-              RAY_LOG(DEBUG) << "SendNumPendingTaskReply: " << WorkerID::FromBinary(data.second.driver_address().worker_id());
+              RAY_LOG(DEBUG) << "SendNumPendingTaskReply: "
+                             << WorkerID::FromBinary(
+                                    data.second.driver_address().worker_id());
               if (!status.ok()) {
                 RAY_LOG(WARNING) << "Failed to get is_running_tasks from core worker: "
                                  << status.ToString();
@@ -301,21 +304,21 @@ std::shared_ptr<rpc::JobConfig> GcsJobManager::GetJobConfig(const JobID &job_id)
 
 void GcsJobManager::OnNodeDead(const NodeID &node_id) {
   RAY_LOG(INFO) << "Node " << node_id
-              << " failed, mark all jobs from this node as finished";
+                << " failed, mark all jobs from this node as finished";
 
-  auto on_done = [this, node_id](
-    const absl::flat_hash_map<JobID, JobTableData> &result) {
-      // If job is not dead and from driver in current node, then mark it as finished
-      for (auto &data : result) {
-        if (!data.second.is_dead() && NodeID::FromBinary(data.second.driver_address().raylet_id()) == node_id) {
-          RAY_LOG(DEBUG) << "Marking job: " << data.first << " as finished";
-          MarkJobAsFinished(data.second, [data](Status status) {
-              if (!status.ok()) {
-                RAY_LOG(WARNING) << "Failed to mark job as finished";
-              }
-          });
-        }
+  auto on_done = [this, node_id](const absl::flat_hash_map<JobID, JobTableData> &result) {
+    // If job is not dead and from driver in current node, then mark it as finished
+    for (auto &data : result) {
+      if (!data.second.is_dead() &&
+          NodeID::FromBinary(data.second.driver_address().raylet_id()) == node_id) {
+        RAY_LOG(DEBUG) << "Marking job: " << data.first << " as finished";
+        MarkJobAsFinished(data.second, [data](Status status) {
+          if (!status.ok()) {
+            RAY_LOG(WARNING) << "Failed to mark job as finished";
+          }
+        });
       }
+    }
   };
 
   // make all jobs in current node to finished

--- a/src/ray/gcs/gcs_server/gcs_job_manager.cc
+++ b/src/ray/gcs/gcs_server/gcs_job_manager.cc
@@ -218,13 +218,13 @@ void GcsJobManager::HandleGetAllJobInfo(rpc::GetAllJobInfoRequest request,
         auto client = core_worker_clients_.GetOrConnect(data.second.driver_address());
         std::unique_ptr<rpc::NumPendingTasksRequest> request(
             new rpc::NumPendingTasksRequest());
-        RAY_LOG(DEBUG) << "SendNumPendingTask: " << worker_id;
+        RAY_LOG(DEBUG) << "Send NumPendingTasksRequest to worker " << worker_id;
         client->NumPendingTasks(
             std::move(request),
             [worker_id, reply, i, num_processed_jobs, try_send_reply](
                 const Status &status,
                 const rpc::NumPendingTasksReply &num_pending_tasks_reply) {
-              RAY_LOG(DEBUG) << "SendNumPendingTaskReply: " << worker_id;
+              RAY_LOG(DEBUG) << "Received NumPendingTasksReply from worker " << worker_id;
               if (!status.ok()) {
                 RAY_LOG(WARNING) << "Failed to get is_running_tasks from core worker: "
                                  << status.ToString();

--- a/src/ray/gcs/gcs_server/gcs_job_manager.cc
+++ b/src/ray/gcs/gcs_server/gcs_job_manager.cc
@@ -204,11 +204,11 @@ void GcsJobManager::HandleGetAllJobInfo(rpc::GetAllJobInfoRequest request,
         job_data_key_to_indices[job_data_key].push_back(i);
       }
 
+      WorkerID worker_id = WorkerID::FromBinary(data.second.driver_address().worker_id());
+
       // If job is not dead, get is_running_tasks from the core worker for the driver.
       if (data.second.is_dead()) {
         reply->mutable_job_info_list(i)->set_is_running_tasks(false);
-        WorkerID worker_id =
-            WorkerID::FromBinary(data.second.driver_address().worker_id());
         core_worker_clients_.Disconnect(worker_id);
         (*num_processed_jobs)++;
         ;
@@ -218,16 +218,12 @@ void GcsJobManager::HandleGetAllJobInfo(rpc::GetAllJobInfoRequest request,
         auto client = core_worker_clients_.GetOrConnect(data.second.driver_address());
         std::unique_ptr<rpc::NumPendingTasksRequest> request(
             new rpc::NumPendingTasksRequest());
-        RAY_LOG(DEBUG) << "SendNumPendingTask: "
-                       << WorkerID::FromBinary(data.second.driver_address().worker_id());
+        RAY_LOG(DEBUG) << "SendNumPendingTask: " << worker_id;
         client->NumPendingTasks(
             std::move(request),
-            [worker_id = data.second.driver_address().worker_id(),
-             reply,
-             i,
-             num_processed_jobs,
-             try_send_reply](const Status &status,
-                             const rpc::NumPendingTasksReply &num_pending_tasks_reply) {
+            [worker_id, reply, i, num_processed_jobs, try_send_reply](
+                const Status &status,
+                const rpc::NumPendingTasksReply &num_pending_tasks_reply) {
               RAY_LOG(DEBUG) << "SendNumPendingTaskReply: " << worker_id;
               if (!status.ok()) {
                 RAY_LOG(WARNING) << "Failed to get is_running_tasks from core worker: "

--- a/src/ray/gcs/gcs_server/gcs_job_manager.cc
+++ b/src/ray/gcs/gcs_server/gcs_job_manager.cc
@@ -222,12 +222,13 @@ void GcsJobManager::HandleGetAllJobInfo(rpc::GetAllJobInfoRequest request,
                        << WorkerID::FromBinary(data.second.driver_address().worker_id());
         client->NumPendingTasks(
             std::move(request),
-            [data, reply, i, num_processed_jobs, try_send_reply](
-                const Status &status,
-                const rpc::NumPendingTasksReply &num_pending_tasks_reply) {
-              RAY_LOG(DEBUG) << "SendNumPendingTaskReply: "
-                             << WorkerID::FromBinary(
-                                    data.second.driver_address().worker_id());
+            [worker_id = data.second.driver_address().worker_id(),
+             reply,
+             i,
+             num_processed_jobs,
+             try_send_reply](const Status &status,
+                             const rpc::NumPendingTasksReply &num_pending_tasks_reply) {
+              RAY_LOG(DEBUG) << "SendNumPendingTaskReply: " << worker_id;
               if (!status.ok()) {
                 RAY_LOG(WARNING) << "Failed to get is_running_tasks from core worker: "
                                  << status.ToString();

--- a/src/ray/gcs/gcs_server/gcs_job_manager.cc
+++ b/src/ray/gcs/gcs_server/gcs_job_manager.cc
@@ -297,5 +297,28 @@ std::shared_ptr<rpc::JobConfig> GcsJobManager::GetJobConfig(const JobID &job_id)
   return it->second;
 }
 
+void GcsJobManager::OnNodeDead(const NodeID &node_id) {
+  RAY_LOG(INFO) << "Node " << node_id
+              << " failed, mark all jobs from this node as finished";
+
+  auto on_done = [this, node_id](
+    const absl::flat_hash_map<JobID, JobTableData> &result) {
+      // If job is not dead and from driver in current node, then mark it as finished
+      for (auto &data : result) {
+        if (!data.second.is_dead() && NodeID::FromBinary(data.second.driver_address().raylet_id()) == node_id) {
+          RAY_LOG(INFO) << "marking job: " << data.first << "as finished";
+          MarkJobAsFinished(data.second, [data](Status status) {
+              if (!status.ok()) {
+                RAY_LOG(WARNING) << "Failed to mark job as finished";
+              }
+          });
+        }
+      }
+  };
+
+  // make all jobs in current node to finished
+  RAY_CHECK_OK(gcs_table_storage_->JobTable().GetAll(on_done));
+}
+
 }  // namespace gcs
 }  // namespace ray

--- a/src/ray/gcs/gcs_server/gcs_job_manager.cc
+++ b/src/ray/gcs/gcs_server/gcs_job_manager.cc
@@ -299,21 +299,21 @@ std::shared_ptr<rpc::JobConfig> GcsJobManager::GetJobConfig(const JobID &job_id)
 
 void GcsJobManager::OnNodeDead(const NodeID &node_id) {
   RAY_LOG(INFO) << "Node " << node_id
-              << " failed, mark all jobs from this node as finished";
+                << " failed, mark all jobs from this node as finished";
 
-  auto on_done = [this, node_id](
-    const absl::flat_hash_map<JobID, JobTableData> &result) {
-      // If job is not dead and from driver in current node, then mark it as finished
-      for (auto &data : result) {
-        if (!data.second.is_dead() && NodeID::FromBinary(data.second.driver_address().raylet_id()) == node_id) {
-          RAY_LOG(INFO) << "marking job: " << data.first << "as finished";
-          MarkJobAsFinished(data.second, [data](Status status) {
-              if (!status.ok()) {
-                RAY_LOG(WARNING) << "Failed to mark job as finished";
-              }
-          });
-        }
+  auto on_done = [this, node_id](const absl::flat_hash_map<JobID, JobTableData> &result) {
+    // If job is not dead and from driver in current node, then mark it as finished
+    for (auto &data : result) {
+      if (!data.second.is_dead() &&
+          NodeID::FromBinary(data.second.driver_address().raylet_id()) == node_id) {
+        RAY_LOG(DEBUG) << "Marking job: " << data.first << " as finished";
+        MarkJobAsFinished(data.second, [data](Status status) {
+          if (!status.ok()) {
+            RAY_LOG(WARNING) << "Failed to mark job as finished";
+          }
+        });
       }
+    }
   };
 
   // make all jobs in current node to finished

--- a/src/ray/gcs/gcs_server/gcs_job_manager.cc
+++ b/src/ray/gcs/gcs_server/gcs_job_manager.cc
@@ -314,7 +314,7 @@ void GcsJobManager::OnNodeDead(const NodeID &node_id) {
         RAY_LOG(DEBUG) << "Marking job: " << data.first << " as finished";
         MarkJobAsFinished(data.second, [data](Status status) {
           if (!status.ok()) {
-            RAY_LOG(WARNING) << "Failed to mark job as finished";
+            RAY_LOG(WARNING) << "Failed to mark job as finished. Status: " << status;
           }
         });
       }

--- a/src/ray/gcs/gcs_server/gcs_job_manager.cc
+++ b/src/ray/gcs/gcs_server/gcs_job_manager.cc
@@ -218,11 +218,13 @@ void GcsJobManager::HandleGetAllJobInfo(rpc::GetAllJobInfoRequest request,
         auto client = core_worker_clients_.GetOrConnect(data.second.driver_address());
         std::unique_ptr<rpc::NumPendingTasksRequest> request(
             new rpc::NumPendingTasksRequest());
+        RAY_LOG(DEBUG) << "SendNumPendingTask: " << WorkerID::FromBinary(data.second.driver_address().worker_id());
         client->NumPendingTasks(
             std::move(request),
             [reply, i, num_processed_jobs, try_send_reply](
                 const Status &status,
                 const rpc::NumPendingTasksReply &num_pending_tasks_reply) {
+              RAY_LOG(DEBUG) << "SendNumPendingTaskReply: " << WorkerID::FromBinary(data.second.driver_address().worker_id());
               if (!status.ok()) {
                 RAY_LOG(WARNING) << "Failed to get is_running_tasks from core worker: "
                                  << status.ToString();
@@ -299,21 +301,21 @@ std::shared_ptr<rpc::JobConfig> GcsJobManager::GetJobConfig(const JobID &job_id)
 
 void GcsJobManager::OnNodeDead(const NodeID &node_id) {
   RAY_LOG(INFO) << "Node " << node_id
-                << " failed, mark all jobs from this node as finished";
+              << " failed, mark all jobs from this node as finished";
 
-  auto on_done = [this, node_id](const absl::flat_hash_map<JobID, JobTableData> &result) {
-    // If job is not dead and from driver in current node, then mark it as finished
-    for (auto &data : result) {
-      if (!data.second.is_dead() &&
-          NodeID::FromBinary(data.second.driver_address().raylet_id()) == node_id) {
-        RAY_LOG(DEBUG) << "Marking job: " << data.first << " as finished";
-        MarkJobAsFinished(data.second, [data](Status status) {
-          if (!status.ok()) {
-            RAY_LOG(WARNING) << "Failed to mark job as finished";
-          }
-        });
+  auto on_done = [this, node_id](
+    const absl::flat_hash_map<JobID, JobTableData> &result) {
+      // If job is not dead and from driver in current node, then mark it as finished
+      for (auto &data : result) {
+        if (!data.second.is_dead() && NodeID::FromBinary(data.second.driver_address().raylet_id()) == node_id) {
+          RAY_LOG(DEBUG) << "Marking job: " << data.first << " as finished";
+          MarkJobAsFinished(data.second, [data](Status status) {
+              if (!status.ok()) {
+                RAY_LOG(WARNING) << "Failed to mark job as finished";
+              }
+          });
+        }
       }
-    }
   };
 
   // make all jobs in current node to finished

--- a/src/ray/gcs/gcs_server/gcs_job_manager.h
+++ b/src/ray/gcs/gcs_server/gcs_job_manager.h
@@ -79,6 +79,12 @@ class GcsJobManager : public rpc::JobInfoHandler {
 
   std::shared_ptr<rpc::JobConfig> GetJobConfig(const JobID &job_id) const;
 
+  /// Handle a node death. This will marks all jobs associated with the
+  /// specified node id as finished.
+  ///
+  /// \param node_id The specified node id.
+  void OnNodeDead(const NodeID &node_id);
+
  private:
   std::shared_ptr<GcsTableStorage> gcs_table_storage_;
   std::shared_ptr<GcsPublisher> gcs_publisher_;

--- a/src/ray/gcs/gcs_server/gcs_server.cc
+++ b/src/ray/gcs/gcs_server/gcs_server.cc
@@ -740,6 +740,7 @@ void GcsServer::InstallEventListeners() {
         gcs_resource_manager_->OnNodeDead(node_id);
         gcs_placement_group_manager_->OnNodeDead(node_id);
         gcs_actor_manager_->OnNodeDead(node_id, node_ip_address);
+        gcs_job_manager_->OnNodeDead(node_id);
         raylet_client_pool_->Disconnect(node_id);
         gcs_healthcheck_manager_->RemoveNode(node_id);
         pubsub_handler_->RemoveSubscriberFrom(node_id.Binary());

--- a/src/ray/gcs/gcs_server/test/gcs_job_manager_test.cc
+++ b/src/ray/gcs/gcs_server/test/gcs_job_manager_test.cc
@@ -594,13 +594,13 @@ TEST_F(GcsJobManagerTest, TestNodeFailure) {
 
   // Check if all job are not dead
   gcs_job_manager.HandleGetAllJobInfo(
-    all_job_info_request,
-    &all_job_info_reply,
-    [&all_job_info_promise](Status, std::function<void()>, std::function<void()>) {
-      all_job_info_promise.set_value(true);
-    });
+      all_job_info_request,
+      &all_job_info_reply,
+      [&all_job_info_promise](Status, std::function<void()>, std::function<void()>) {
+        all_job_info_promise.set_value(true);
+      });
   all_job_info_promise.get_future().get();
-  for(auto job_info: all_job_info_reply.job_info_list()){
+  for (auto job_info : all_job_info_reply.job_info_list()) {
     ASSERT_TRUE(!job_info.is_dead());
   }
 
@@ -611,27 +611,27 @@ TEST_F(GcsJobManagerTest, TestNodeFailure) {
 
   // Test get all jobs with limit larger than the number of jobs.
   auto condition = [&gcs_job_manager, node_id]() -> bool {
-      rpc::GetAllJobInfoRequest all_job_info_request2;
-      rpc::GetAllJobInfoReply all_job_info_reply2;
-      std::promise<bool> all_job_info_promise2;
-      gcs_job_manager.HandleGetAllJobInfo(
-          all_job_info_request2,
-          &all_job_info_reply2,
-          [&all_job_info_promise2](Status, std::function<void()>, std::function<void()>) {
-            all_job_info_promise2.set_value(true);
-          });
-      all_job_info_promise2.get_future().get();
+    rpc::GetAllJobInfoRequest all_job_info_request2;
+    rpc::GetAllJobInfoReply all_job_info_reply2;
+    std::promise<bool> all_job_info_promise2;
+    gcs_job_manager.HandleGetAllJobInfo(
+        all_job_info_request2,
+        &all_job_info_reply2,
+        [&all_job_info_promise2](Status, std::function<void()>, std::function<void()>) {
+          all_job_info_promise2.set_value(true);
+        });
+    all_job_info_promise2.get_future().get();
 
-      auto job_info1 = all_job_info_reply2.job_info_list().Get(0);
-      auto job_info2 = all_job_info_reply2.job_info_list().Get(1);
+    auto job_info1 = all_job_info_reply2.job_info_list().Get(0);
+    auto job_info2 = all_job_info_reply2.job_info_list().Get(1);
 
-      bool job_condition = true;
-      // job1 from the current node should dead, while job2 is still alive
-      for(auto job_info: all_job_info_reply2.job_info_list()){
-        auto job_node_id = NodeID::FromBinary(job_info.driver_address().raylet_id());
-        job_condition = job_condition && (job_info.is_dead() == (job_node_id == node_id));
-      }
-      return job_condition;
+    bool job_condition = true;
+    // job1 from the current node should dead, while job2 is still alive
+    for (auto job_info : all_job_info_reply2.job_info_list()) {
+      auto job_node_id = NodeID::FromBinary(job_info.driver_address().raylet_id());
+      job_condition = job_condition && (job_info.is_dead() == (job_node_id == node_id));
+    }
+    return job_condition;
   };
 
   EXPECT_TRUE(WaitForCondition(condition, 2000));

--- a/src/ray/gcs/gcs_server/test/gcs_job_manager_test.cc
+++ b/src/ray/gcs/gcs_server/test/gcs_job_manager_test.cc
@@ -607,37 +607,34 @@ TEST_F(GcsJobManagerTest, TestNodeFailure) {
   // Remove node and then check that the job is dead.
   auto address = all_job_info_reply.job_info_list().Get(0).driver_address();
   auto node_id = NodeID::FromBinary(address.raylet_id());
-  
-  std::promise<bool> on_node_dead_promise;
-  io_service_.post(
-    [&gcs_job_manager, node_id, &on_node_dead_promise]() {
-      gcs_job_manager.OnNodeDead(node_id);
-      on_node_dead_promise.set_value(true);
-    },
-    "test");
-  ASSERT_TRUE(on_node_dead_promise.get_future().get());
+  gcs_job_manager.OnNodeDead(node_id);
 
   // Test get all jobs with limit larger than the number of jobs.
-  rpc::GetAllJobInfoRequest all_job_info_request2;
-  rpc::GetAllJobInfoReply all_job_info_reply2;
-  std::promise<bool> all_job_info_promise2;
-  gcs_job_manager.HandleGetAllJobInfo(
-      all_job_info_request2,
-      &all_job_info_reply2,
-      [&all_job_info_promise2](Status, std::function<void()>, std::function<void()>) {
-        all_job_info_promise2.set_value(true);
-      });
-  all_job_info_promise2.get_future().get();
+  auto condition = [&gcs_job_manager, node_id]() -> bool {
+      rpc::GetAllJobInfoRequest all_job_info_request2;
+      rpc::GetAllJobInfoReply all_job_info_reply2;
+      std::promise<bool> all_job_info_promise2;
+      gcs_job_manager.HandleGetAllJobInfo(
+          all_job_info_request2,
+          &all_job_info_reply2,
+          [&all_job_info_promise2](Status, std::function<void()>, std::function<void()>) {
+            all_job_info_promise2.set_value(true);
+          });
+      all_job_info_promise2.get_future().get();
 
-  ASSERT_EQ(all_job_info_reply2.job_info_list().size(), 2);
-  auto job_info1 = all_job_info_reply2.job_info_list().Get(0);
-  auto job_info2 = all_job_info_reply2.job_info_list().Get(1);
+      auto job_info1 = all_job_info_reply2.job_info_list().Get(0);
+      auto job_info2 = all_job_info_reply2.job_info_list().Get(1);
 
-  // job1 from the current node should dead, while job2 is still alive
-  for(auto job_info: all_job_info_reply2.job_info_list()){
-    auto job_node_id = NodeID::FromBinary(job_info.driver_address().raylet_id());
-    ASSERT_EQ(job_info.is_dead(), job_node_id == node_id);
-  }
+      bool job_condition = true;
+      // job1 from the current node should dead, while job2 is still alive
+      for(auto job_info: all_job_info_reply2.job_info_list()){
+        auto job_node_id = NodeID::FromBinary(job_info.driver_address().raylet_id());
+        job_condition = job_condition && (job_info.is_dead() == (job_node_id == node_id));
+      }
+      return job_condition;
+  };
+
+  EXPECT_TRUE(WaitForCondition(condition, 2000));
 }
 
 int main(int argc, char **argv) {

--- a/src/ray/gcs/gcs_server/test/gcs_job_manager_test.cc
+++ b/src/ray/gcs/gcs_server/test/gcs_job_manager_test.cc
@@ -553,6 +553,93 @@ TEST_F(GcsJobManagerTest, TestPreserveDriverInfo) {
   ASSERT_EQ(data.driver_pid(), 8264);
 }
 
+TEST_F(GcsJobManagerTest, TestNodeFailure) {
+  gcs::GcsJobManager gcs_job_manager(gcs_table_storage_,
+                                     gcs_publisher_,
+                                     runtime_env_manager_,
+                                     *function_manager_,
+                                     *fake_kv_,
+                                     client_factory_);
+
+  auto job_id1 = JobID::FromInt(1);
+  auto job_id2 = JobID::FromInt(2);
+  gcs::GcsInitData gcs_init_data(gcs_table_storage_);
+  gcs_job_manager.Initialize(/*init_data=*/gcs_init_data);
+
+  rpc::AddJobReply empty_reply;
+  std::promise<bool> promise1;
+  std::promise<bool> promise2;
+
+  auto add_job_request1 = Mocker::GenAddJobRequest(job_id1, "namespace_1");
+  gcs_job_manager.HandleAddJob(
+      *add_job_request1,
+      &empty_reply,
+      [&promise1](Status, std::function<void()>, std::function<void()>) {
+        promise1.set_value(true);
+      });
+  promise1.get_future().get();
+
+  auto add_job_request2 = Mocker::GenAddJobRequest(job_id2, "namespace_2");
+  gcs_job_manager.HandleAddJob(
+      *add_job_request2,
+      &empty_reply,
+      [&promise2](Status, std::function<void()>, std::function<void()>) {
+        promise2.set_value(true);
+      });
+  promise2.get_future().get();
+
+  rpc::GetAllJobInfoRequest all_job_info_request;
+  rpc::GetAllJobInfoReply all_job_info_reply;
+  std::promise<bool> all_job_info_promise;
+
+  // Check if all job are not dead
+  gcs_job_manager.HandleGetAllJobInfo(
+    all_job_info_request,
+    &all_job_info_reply,
+    [&all_job_info_promise](Status, std::function<void()>, std::function<void()>) {
+      all_job_info_promise.set_value(true);
+    });
+  all_job_info_promise.get_future().get();
+  for(auto job_info: all_job_info_reply.job_info_list()){
+    ASSERT_TRUE(!job_info.is_dead());
+  }
+
+  // Remove node and then check that the job is dead.
+  auto address = all_job_info_reply.job_info_list().Get(0).driver_address();
+  auto node_id = NodeID::FromBinary(address.raylet_id());
+  
+  std::promise<bool> on_node_dead_promise;
+  io_service_.post(
+    [&gcs_job_manager, node_id, &on_node_dead_promise]() {
+      gcs_job_manager.OnNodeDead(node_id);
+      on_node_dead_promise.set_value(true);
+    },
+    "test");
+  ASSERT_TRUE(on_node_dead_promise.get_future().get());
+
+  // Test get all jobs with limit larger than the number of jobs.
+  rpc::GetAllJobInfoRequest all_job_info_request2;
+  rpc::GetAllJobInfoReply all_job_info_reply2;
+  std::promise<bool> all_job_info_promise2;
+  gcs_job_manager.HandleGetAllJobInfo(
+      all_job_info_request2,
+      &all_job_info_reply2,
+      [&all_job_info_promise2](Status, std::function<void()>, std::function<void()>) {
+        all_job_info_promise2.set_value(true);
+      });
+  all_job_info_promise2.get_future().get();
+
+  ASSERT_EQ(all_job_info_reply2.job_info_list().size(), 2);
+  auto job_info1 = all_job_info_reply2.job_info_list().Get(0);
+  auto job_info2 = all_job_info_reply2.job_info_list().Get(1);
+
+  // job1 from the current node should dead, while job2 is still alive
+  for(auto job_info: all_job_info_reply2.job_info_list()){
+    auto job_node_id = NodeID::FromBinary(job_info.driver_address().raylet_id());
+    ASSERT_EQ(job_info.is_dead(), job_node_id == node_id);
+  }
+}
+
 int main(int argc, char **argv) {
   ::testing::InitGoogleTest(&argc, argv);
   return RUN_ALL_TESTS();

--- a/src/ray/gcs/gcs_server/test/gcs_job_manager_test.cc
+++ b/src/ray/gcs/gcs_server/test/gcs_job_manager_test.cc
@@ -609,7 +609,7 @@ TEST_F(GcsJobManagerTest, TestNodeFailure) {
   auto node_id = NodeID::FromBinary(address.raylet_id());
   gcs_job_manager.OnNodeDead(node_id);
 
-  // Test get all jobs with limit larger than the number of jobs.
+  // Test get all jobs and check if killed node jobs marked as finished
   auto condition = [&gcs_job_manager, node_id]() -> bool {
     rpc::GetAllJobInfoRequest all_job_info_request2;
     rpc::GetAllJobInfoReply all_job_info_reply2;
@@ -621,9 +621,6 @@ TEST_F(GcsJobManagerTest, TestNodeFailure) {
           all_job_info_promise2.set_value(true);
         });
     all_job_info_promise2.get_future().get();
-
-    auto job_info1 = all_job_info_reply2.job_info_list().Get(0);
-    auto job_info2 = all_job_info_reply2.job_info_list().Get(1);
 
     bool job_condition = true;
     // job1 from the current node should dead, while job2 is still alive

